File: src/frontend/src/lib/trade/tradeRatio.ts
// Mirrors src/backend/src/Bjarnoy.Domain/Trade/TradeRatio.cs exactly (issue
// #46's design brief): a player may offer at most `DEFAULT_MAX_RATIO` units
// of one resource for one unit of another on the open market, or
// `GUILD_MAX_RATIO` between guild mates. Kept here as a standalone module
// (rather than inlined in TradePanel.vue) so both the live-mode instant
// client-side check (TradePanel) and the demo-mode simulation
// (WorldModel.postTradeOffer) validate against the exact same rule instead
// of two hand-copied versions drifting apart.

/** Max ratio (offered:requested or requested:offered) on the open market. */
export const DEFAULT_MAX_RATIO = 2;

/** Max ratio between guild mates. */
export const GUILD_MAX_RATIO = 8;

export type TradeRatioRejection = 'ZeroAmount' | 'SameResource' | 'RatioExceeded' | null;

/**
 * Validates offered/requested amounts and resources against the ratio
 * corridor, exactly as `TradeRatio.Validate` does server-side. Returns the
 * rejection reason (matching the backend's `TradeRejection` names) or null
 * if the trade is within bounds. Does not check resource stock/carts/
 * longhouse level — those are server-only (or, in demo mode,
 * WorldModel-only) checks that need more context than a ratio check does.
 */
export function validateTradeRatio(
  offeredResource: string,
  offeredAmount: number,
  requestedResource: string,
  requestedAmount: number,
  guildOnly: boolean,
): TradeRatioRejection {
  if (offeredAmount <= 0 || requestedAmount <= 0) return 'ZeroAmount';
  if (offeredResource === requestedResource) return 'SameResource';

  const maxRatio = guildOnly ? GUILD_MAX_RATIO : DEFAULT_MAX_RATIO;
  // Symmetric corridor: at 1:2 you may post 400 for 200 or 200 for 400, but
  // never 500 for 200 in either direction.
  if (offeredAmount > maxRatio * requestedAmount || requestedAmount > maxRatio * offeredAmount) {
    return 'RatioExceeded';
  }
  return null;
}

File: src/backend/src/Bjarnoy.Domain/Trade/TradeRatio.cs
namespace Bjarnoy.Domain.Trade;

/// <summary>
/// The trade-ratio rule from the design brief (issue #46): a player may offer
/// at most <see cref="DefaultMaxRatio"/> units of one resource for one unit of
/// another on the open market, or <see cref="GuildMaxRatio"/> between guild
/// mates.
/// </summary>
/// <remarks>
/// A pure function with no guild lookup of its own — <paramref name="isGuildTrade"/>
/// (see <see cref="Validate"/>) is a fact the caller resolves and hands in.
/// There is no Guild domain type yet; the rule will not need to change when
/// one exists; only how the bool is produced will.
/// </remarks>
public static class TradeRatio
{
    /// <summary>Max ratio (offered:requested or requested:offered) on the open market.</summary>
    public const int DefaultMaxRatio = 2;

    /// <summary>Max ratio between guild mates.</summary>
    public const int GuildMaxRatio = 8;

    /// <summary>
    /// Whether an offer of <paramref name="offeredAmount"/> <paramref name="offeredResource"/>
    /// for <paramref name="requestedAmount"/> <paramref name="requestedResource"/> is within the
    /// allowed ratio corridor.
    /// </summary>
    public static TradeRejection Validate(
        TradeResource offeredResource,
        double offeredAmount,
        TradeResource requestedResource,
        double requestedAmount,
        bool isGuildTrade)
    {
        if (offeredAmount <= 0 || requestedAmount <= 0)
        {
            return TradeRejection.ZeroAmount;
        }

        if (offeredResource == requestedResource)
        {
            return TradeRejection.SameResource;
        }

        var maxRatio = isGuildTrade ? GuildMaxRatio : DefaultMaxRatio;

        // Symmetric corridor: at 1:2 you may post 400 for 200 or 200 for 400,
        // but never 500 for 200 in either direction.
        if (offeredAmount > maxRatio * requestedAmount || requestedAmount > maxRatio * offeredAmount)
        {
            return TradeRejection.RatioExceeded;
        }

        return TradeRejection.None;
    }
}
